### PR TITLE
Reimplement "FCOS: Fully Convolutional One-Stage Object Detection"

### PR DESCRIPTION
**Pull request by yhcao6**
The code refers to https://github.com/tianzhi0549/FCOS

Performance are measured on 4-v100 with 4 images per GPU

|              | paper | this repo |
|--------------|-------|-----------|
| FCOS R50 FPN | 36.6  | 36.6      |

| Backbone | Style | Lr schd | Mem (GB) | Train time (s/iter) | Inf time (fps) | box AP |
|----------|-------|---------|----------|---------------------|----------------|--------|
| R-50-FPN | caffe | 1x      | 7.5      | 0.403               | 13.5           | 36.6   |

### CODE REVIEW

File: mmdet/models/anchor_heads/fcos_head.py [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+                 num_classes,
+                 in_channels,
+                 feat_channels=256,
+                 stacked_convs=4,
+                 strides=(4, 8, 16, 32, 64),
+                 regress_ranges=((-1, 64), (64, 128), (128, 256), (256, 512),
+                                 (512, INF))):
+        super(FCOSHead, self).__init__()
+
+        self.num_classes = num_classes - 1
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: This is confusing, just use another variable to do so.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+    def _init_layers(self):
+        self.cls_layers = nn.ModuleList()
+        self.reg_layers = nn.ModuleList()
+        for i in range(self.stacked_convs):
+            chn = self.in_channels if i == 0 else self.feat_channels
+            self.cls_layers.append(
+                nn.Conv2d(chn, self.feat_channels, 3, stride=1, padding=1))
+            self.cls_layers.append(nn.GroupNorm(32, self.feat_channels))
+            self.cls_layers.append(nn.ReLU(inplace=False))
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: Use `ConvModule` instead of 3 layers.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+            self.reg_layers.append(
+                nn.Conv2d(chn, self.feat_channels, 3, stride=1, padding=1))
+            self.reg_layers.append(nn.GroupNorm(32, self.feat_channels))
+            self.reg_layers.append(nn.ReLU(inplace=False))
+        self.fcos_cls = nn.Conv2d(self.feat_channels, self.num_classes, 3, 1,
+                                  1)
+        self.fcos_reg = nn.Conv2d(self.feat_channels, 4, 3, 1, 1)
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: Explicitly specify `padding`, `stride` and `dilation` to make it more readable.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+        self.in_channels = in_channels
+        self.feat_channels = feat_channels
+        self.stacked_convs = stacked_convs
+        self.strides = strides
+        self.regress_ranges = regress_ranges
+
+        self._init_layers()
+
+    def _init_layers(self):
+        self.cls_layers = nn.ModuleList()
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: Make the names consistent with other heads.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        reg_feat = x
+
+        for cls_layer in self.cls_layers:
+            cls_feat = cls_layer(cls_feat)
+        cls_score = self.fcos_cls(cls_feat)
+        centerness = self.fcos_centerness(cls_feat)
+
+        for reg_layer in self.reg_layers:
+            reg_feat = reg_layer(reg_feat)
+        bbox_pred = torch.exp(scale(self.fcos_reg(reg_feat)))
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: Use `.exp()` to make it short.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            mlvl_bboxes,
+            mlvl_scores,
+            cfg.score_thr,
+            cfg.nms,
+            cfg.max_per_img,
+            score_cofficient=mlvl_centerness)
+        return det_bboxes, det_labels
+
+    def get_centers(self, feat_sizes, strides, dtype, device):
+        return multi_apply(
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: There is no need to use `multi_apply` and `get_centers_single`.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+        det_bboxes, det_labels = multiclass_nms(
+            mlvl_bboxes,
+            mlvl_scores,
+            cfg.score_thr,
+            cfg.nms,
+            cfg.max_per_img,
+            score_cofficient=mlvl_centerness)
+        return det_bboxes, det_labels
+
+    def get_centers(self, feat_sizes, strides, dtype, device):
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: The argument `strides` can be removed.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+             bbox_preds,
+             centernesses,
+             gt_bboxes,
+             gt_labels,
+             img_metas,
+             cfg,
+             gt_bboxes_ignore=None):
+        featmap_sizes = [featmap.size()[-2:] for featmap in cls_scores]
+        all_level_centers = self.get_centers(featmap_sizes, self.strides,
+                                             cls_scores[0].dtype,
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: `bbox_preds[0].dtype` is more appropriate.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+    def loss(self,
+             cls_scores,
+             bbox_preds,
+             centernesses,
+             gt_bboxes,
+             gt_labels,
+             img_metas,
+             cfg,
+             gt_bboxes_ignore=None):
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: Add an assertion here to check the inputs.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from mmcv.cnn import normal_init
+
+from mmdet.core import (centerness_target, fcos_target, sigmoid_focal_loss,
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: Implement `centerness_target` and `fcos_target` as methods of `FCOSHead`. There will be a refactoring for `anchor_target`.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+            ]
+            bbox_pred_list = [
+                bbox_preds[i][img_id].detach() for i in range(num_levels)
+            ]
+            centerness_pred_list = [
+                centernesses[i][img_id].detach() for i in range(num_levels)
+            ]
+            img_shape = img_metas[img_id]['img_shape']
+            scale_factor = img_metas[img_id]['scale_factor']
+            proposals = self.get_bboxes_single(
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: This is not proposals.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+
+            x1 = centers[:, 0] - bbox_pred[:, 0]
+            y1 = centers[:, 1] - bbox_pred[:, 1]
+            x2 = centers[:, 0] + bbox_pred[:, 2]
+            y2 = centers[:, 1] + bbox_pred[:, 3]
+            x1 = x1.clamp(min=0, max=img_shape[1] - 1)
+            y1 = y1.clamp(min=0, max=img_shape[0] - 1)
+            x2 = x2.clamp(min=0, max=img_shape[1] - 1)
+            y2 = y2.clamp(min=0, max=img_shape[0] - 1)
+            bboxes = torch.stack([x1, y1, x2, y2], -1)
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: add a method `distance2bbox()` to `core/bbox/transforms.py`.

File: mmdet/models/necks/fpn.py [hunk 13]
```diff
@@ -93,7 +95,10 @@ def __init__(self,
     def init_weights(self):
         for m in self.modules():
             if isinstance(m, nn.Conv2d):
-                xavier_init(m, distribution='uniform')
+                if self.relu_extra_convs:
+                    caffe2_xavier_init(m)
```
Comment (hellock) [inline mmdet/models/necks/fpn.py]: There is no logical relation between these two statements.

File: mmdet/core/utils/initialize_utils.py [hunk 14]
```diff
@@ -0,0 +1,36 @@
+import torch.optim as optim
+
+
+def caffe2_initialize(model, cfg):
```
Comment (hellock) [inline mmdet/core/utils/initialize_utils.py]: Using different lr and decay for weight and bias is not specific to caffe2, which has been adopted long ago. The naming is misleading.

File: mmdet/core/loss/losses.py [hunk 15]
```diff
@@ -127,3 +127,36 @@ def _expand_binary_labels(labels, label_weights, label_channels):
     bin_label_weights = label_weights.view(-1, 1).expand(
         label_weights.size(0), label_channels)
     return bin_labels, bin_label_weights
+
+
+def iou_loss(pred, target, reduction='mean'):
```
Comment (hellock) [inline mmdet/core/loss/losses.py]: Please refer to https://github.com/open-mmlab/mmdetection/blob/master/mmdet/core/bbox/geometry.py#L4 for more efficient implementation.

File: mmdet/core/post_processing/bbox_nms.py [hunk 16]
```diff
@@ -34,10 +36,13 @@ def multiclass_nms(multi_bboxes, multi_scores, score_thr, nms_cfg, max_num=-1):
         else:
             _bboxes = multi_bboxes[cls_inds, i * 4:(i + 1) * 4]
         _scores = multi_scores[cls_inds, i]
+        if score_cofficient is not None:
+            _scores *= score_cofficient[cls_inds]
```
Comment (hellock) [inline mmdet/core/post_processing/bbox_nms.py]: This should be done before this method is called (outside of the method).

File: mmdet/models/anchor_heads/fcos_head.py [hunk 17]
```diff
@@ -0,0 +1,356 @@
...
+                ConvModule(
+                    chn,
+                    self.feat_channels,
+                    3,
+                    stride=1,
+                    padding=1,
+                    normalize=self.normalize,
+                    bias=self.normalize is None))
+        self.fcos_cls = nn.Conv2d(
+            self.feat_channels, self.cls_out_channels, 3, stride=1, padding=1)
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: `stride=1` can be omitted.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 18]
```diff
@@ -0,0 +1,356 @@
...
+        self.fcos_reg = nn.Conv2d(
+            self.feat_channels, 4, 3, stride=1, padding=1)
+        self.fcos_centerness = nn.Conv2d(
+            self.feat_channels, 1, 3, stride=1, padding=1)
+
+        self.scales = nn.ModuleList([Scale(1.0) for _ in range(5)])
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: The hardcoded number `5` should be replaced.

File: mmdet/models/anchor_heads/fcos_head.py [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+from mmdet.core import (sigmoid_focal_loss, iou_loss, multi_apply,
+                        multiclass_nms, distance2bbox)
+from ..registry import HEADS
+from ..utils import bias_init_with_prob, Scale, ConvModule
+
+INF = 1e8
+
+
+@HEADS.register_module
+class FCOSHead(nn.Module):
```
Comment (hellock) [inline mmdet/models/anchor_heads/fcos_head.py]: There is few comments for this class.

File: mmdet/models/necks/fpn.py [hunk 20]
```diff
@@ -131,5 +138,8 @@ def forward(self, inputs):
                     outs.append(self.fpn_convs[used_backbone_levels](outs[-1]))
                 for i in range(used_backbone_levels + 1, self.num_outs):
                     # BUG: we should add relu before each extra conv
```
Comment (hellock) [inline mmdet/models/necks/fpn.py]: Remove this line.

File: mmdet/models/necks/fpn.py [hunk 21]
```diff
@@ -19,7 +19,9 @@ def __init__(self,
                  extra_convs_on_inputs=True,
                  conv_cfg=None,
                  normalize=None,
-                 activation=None):
+                 activation=None,
+                 relu_extra_convs=False,
```
Comment (hellock) [inline mmdet/models/necks/fpn.py]: May be better to rename it to `relu_before_extra_convs`. This argument can be moved after `extra_convs_on_inputs`.

File: mmdet/core/utils/initialize_utils.py (hunk 14 above)
Comment (hellock) [reply mmdet/core/utils/initialize_utils.py]: Add three fileds `bias_lr_mult`, `bias_decay_mult`, `norm_layer_weight_decay` in config files.

File: mmdet/core/utils/initialize_utils.py [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    optimizer_cfg = cfg.copy()
+    optimizer_type = optimizer_cfg.pop('type')
+
+    params = []
+    base_lr = optimizer_cfg.lr
+    base_weight_decay = optimizer_cfg.weight_decay
+    for k, v in model.named_parameters():
+        if not v.requires_grad:
+            continue
+        if 'bias' in k:
```
Comment (hellock) [inline mmdet/core/utils/initialize_utils.py]: `k.endswith('.bias')`

File: configs/fcos/fcos_mstrain_640_800_r101_caffe_fpn_gn_2x_4gpu.py [hunk 23]
```diff
@@ -0,0 +1,124 @@
...
+        dict(type='TextLoggerHook'),
+        # dict(type='TensorboardLoggerHook')
+    ])
+# yapf:enable
+# runtime settings
+total_epochs = 24
+device_ids = range(4)
+dist_params = dict(backend='nccl')
+log_level = 'INFO'
+work_dir = './work_dirs/fcos_r101_fpn_2x'
```
Comment (hellock) [inline configs/fcos/fcos_mstrain_640_800_r101_caffe_fpn_gn_2x_4gpu.py]: Make the work_dir consistent with the config filename.
